Accept header proposal

Half-Shot opened the pull request.

ara4n commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,71 @@
+
+## Tradeoffs
+
+I cannot think of another way to do this that brings in a benefit over this.
```
I've been thinking about this a bit too, and wondered whether we should be instead spending time specifying alternative transports rather than alternative encodings and how we would best upgrade from the HTTPS+JSON baseline to something like CoAP+CBOR.  This feels like a good way to negotiate HTTPS+CBOR rather than HTTPS+JSON, but I wonder how much of a win CBOR alone actually is in this context: when i've benchmarked it in the past it wasn't a massive improvement over JSON.  For instance, my test initial sync is 4.8MB of JSON or 3.9MB of CBOR (so only a 20% win).

So, at the risk of opening the pandorica: i wonder if we should consider solving this problem  by providing a way to negotiate both transport & encoding at the same time, rather than special-casing it just for custom encodings over HTTP.  This could be done by upgrading from plain HTTPS+JSON to sexier protocols (e.g. CBOR+CoAP) - perhaps by mimicking how HTTP/3 upgrades are done (which is a pretty similar problem?)  Or perhaps by advertising what transports your server supports in .well-known?

Half-Shot replied:
Yep. The proposal was intended to be a 'quick win' by making use of existing negotiation standards rather than pushing the boat out to support transports. In terms of savings, I also found only a 20% win with CBOR (with an initial sync test), but please remember this is a proposal to negotiate rather than for CBOR.

There might be situations where HTTP is the only method available to the client/server (i.e. horrible corporate firewalls blocking everything under the sun) and this becomes one of the best options available to a client.

I'm not totally convinced we can't perceive them as different things and have well-known/SRV records point out transports supported by your home server, and then have the transport negotiate an encoding separately rather than reinventing half the wheel.

Half-Shot replied:
(I've added transport negotiation as an alternative solution however)

bwindels replied:
+1 that tying the encoding to the transport might make it harder to just get the quick encoding win where a non-TCP-on-port-443 transport can't be deployed...

ara4n replied:
fair enough.

maxidor commented:
Is this any different from what is already in HTTP RFCs in terms of content negotiation?

uhoreg commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,77 @@
+and decide based on the rules provided in the RFC what is the best encoding they can support.
+
+If a encoding could not be found that satifies both parties, the server should default to JSON. This is
+due to historical reasons as Matrix has presumed JSON support in all implementations. Another benefit to this
```
IMHO, if the `Accept` header is present and no common encoding is found, then it should be an error, whereas if no `Accept` header is present, then it should default to JSON.

uhoreg replied:
FWIW, the RFC says "A request without any Accept header field implies that the user agent will accept any media type in response."

bwindels replied:
Would this then allow to use the standard `406` as a response code without a JSON body?

Half-Shot commented:
>Is this any different from what is already in HTTP RFCs in terms of content negotiation?

My apologies, I meant to answer this one: No, it's following the same model of the client specifying in `Accept` what they will accept and the server either responding with the best encoding it can offer, or a 406 if it cannot. Once that initial response has happened, the client will then use the encoding that the server has responded with in `Content-Type`.

ara4n commented:
hm, i wonder if it'd be easier to spec this as "we follow HTTP content negotiation as per https://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html, but we mandate a 406 if none of the accept headers are supported"?

turt2live commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,83 @@
+# MSC 1740: Using the Accept header to select an encoding
+
+Matrix has always aimed to support many types of encoding However, in practice the only supported option
```
nit: `However` shouldn't be capitalized or should be preceded by a stop.

Half-Shot commented:
> hm, i wonder if it'd be easier to spec this as "we follow HTTP content negotiation as per https://www.w3.org/Protocols/rfc2616/rfc2616-sec12.html, but we mandate a 406 if none of the accept headers are supported"?

Absolutely, though do we care about what errors we send back to the clients if any on a 406 (or a 415 for the client sending a unacceptable content-type)?

I think this would be fine as a small section in the spec.

maxidor commented:
> Is this any different from what is already in HTTP RFCs in terms of content negotiation?

For follow-up: yes it is different and does not try to strictly stick to HTTP RFCs.

turt2live commented on proposals/1740-accept-header.md:
````diff
@@ -0,0 +1,83 @@
+```json
+{
+  "errcode": "M_CONTENT_TYPE_NOT_SUPPORTED",
+  "error": "..error message left to the discretion of the implementation.."
````
should also include a list of supported types, to prevent clients from guessing forever. According to HTTP 406's spec, we *SHOULD* supply this information anyways in the payload.

turt2live commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,83 @@
+### Clients
+
+Clients SHOULD supply `Accept` to all requests they make, and set `Content-Type` to the encoding
+they intend to use. If a client doesn't supply an `Accept` header, then JSON must be presumed acceptable
```
nit: s/JSON/`application/json` (use the actual mime type in the spec)

Half-Shot commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,85 @@
+An appropriate error may be displayed to the user.
+
+It is suggested (but not required), that clients first request `/_matrix/client/versions`  from the
+sever to ensure that `Accept` is acceptable and by doing so negotiate an appropriate encoding.
```
Agreed, this section doesn't really need saying.

Half-Shot commented:
@KitsuneRal That's indeed the idea.

jevolk commented:
Can anyone explain how this isn't plagiarizing the HTTP specification? What are the boundaries for what's relevant to be included for Matrix? Can I re-specify TLS? What about TCP? Should I instead just submit proposals for every HTTP header [Construct](https://github.com/matrix-construct/construct) makes use of?

Where does it end? It seems as if you guys are just submitting things make it look like you're submitting things when someone lists all your contributions. The reviewers of this PR are all holding each other's  here. This is specification fraud.

Who is going to fall for this? @ara4n you should set some standards here, it reflects poorly on you.

ara4n commented:
@jevolk nobody is suggesting copying any of the content nego spec from HTTP into Matrix. The intention, as per https://github.com/matrix-org/matrix-doc/pull/1740#issuecomment-444062734, is to just confirm that Matrix follows HTTP's content nego (spelling out explicitly how nego failures should be handled for Matrix's use case).

Please stop screeching.

jevolk commented:
Consider the [Mastodon specification](https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md) in contrast. Mastodon is a mature and fully functional restful application running on HTTP -- no different than Matrix in this regard. It has an established and diverse set of implementations from several parties. I don't see this kind of nonsense there. Their specification is kept clean and simple, probably because they want people to respect it and actually be able to implement it. Perhaps you guys don't want that.

ara4n commented:
*sighs and blocks*.

Just in case anyone else has the misfortune of reading this: just because you have a point which has some technical merits doesn't mean you can scream abuse at other contributors and act like a complete jerk.  We will block folks from participating in the project if they act in this way, as we believe it seriously discourages the vast majority of current and future contributors from wanting to participate - and that negative impact on the project massively outweighs any positive value.

Luckily, at least 3 other people have made your technical point already here, so nothing of value was lost.

Half-Shot commented:
Would it be possible to get a show of approvals/needs work from folks who have commented so I can see where we are?

bwindels commented on proposals/1740-accept-header.md:
````diff
@@ -0,0 +1,82 @@
+```json
+{
+  "errcode": "M_NOT_ACCEPTABLE",
+  "accepts": "text/html;q=0.2 application/json"
````
Nit: missing comma

turt2live replied:
two of them: one at the end of the line and one after `q=0.2`

turt2live commented on proposals/1740-accept-header.md:
````diff
@@ -0,0 +1,82 @@
+```json
+{
+  "errcode": "M_CONTENT_TYPE_NOT_SUPPORTED",
+  "error": "..error message left to the discretion of the implementation.."
````
the 415 should also probably have an `accepts` key like the 406.

Half-Shot commented:
Added key to the `415` error. I have changed the format for `accepts` for the reasons given in the document. Namely it doesn't make much sense for the server to be telling us wild card results, or quality rankings, and finally it seems harsh to make the client parse the response like that.

uhoreg commented on proposals/1740-accept-header.md:
````diff
@@ -0,0 +1,87 @@
+```
+where accepts is the set of acceptable encodings. This is NOT in `Accept` format because:
+
+1) Servers should specify a quality or ordering, as the negotiations are client led.
````
s/should/should not/ ?

richvdh commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,87 @@
+If the rules of `Accept` fail (a satisfactory encoding could not be picked), the server SHOULD send a 
+`HTTP 406 Not Acceptable`.
+
+If 406 is given, the server MUST also specify the set of acceptable formats described as:
```
I'm somewhat wondering why we bother. Presumably the client has listed everything it supports, so knowledge of what the server supports doesn't really help anyone.

turt2live replied:
the HTTP spec recommends that servers send back what it does support, although doesn't specify how. My interpretation of the intention is one for debugging purposes: it can help quickly identify a problem where the server doesn't support the encoding. In the scenario of someone coming to an issue tracker saying "the client doesn't work", the author could see that it's a matter of unsupported formats rather than some other bug.

richvdh replied:
@turt2live I think you're probably right, but in that case it seems overkill to tightly specify a response format.

turt2live replied:
well, the argument I made for using the standard error response (even though we don't have to) is that clients may fall back to parsing errors as JSON regardless of transport. When we use the standard error response, it is likely that the message we're trying to convey is left in-tact. 

The artifact of using JSON is that `accepts` gets strictly specified.

richvdh replied:
so for us to get into the situation of returning a 406, that implies that either the client doesn't support JSON (in which case, returning a JSON error seems pointless), or the server doesn't support JSON (in which case, having it return this one error as a JSON dict seems a bit strange). Either way I don't quite follow why tightly specifying a response format buys us anything here above what the HTTP RFCs already say.

Half-Shot replied:
I believe we should be erroring with a 406 and 415 as appropriate, but I don't really care if we error with a JSON value or something plaintext or as you say, anything the server wishes.

turt2live replied:
I'm fine to leave the entire body undefined and up to the implementation, although if we put anything in the spec it should be JSON imho. I just don't see the value in intentionally changing away from JSON for this endpoint.

maxidor replied:
Just as a reminder, as per [the spec itself](https://matrix.org/docs/spec/client_server/unstable.html#api-standards):
> The mandatory baseline for client-server communication in Matrix is exchanging JSON objects over HTTP APIs.

> For the default HTTP transport, all API calls use a Content-Type of application/json.

> Any errors which occur at the Matrix API level MUST return a "standard error response".

If the MSC could be written keeping those in mind, that would be great.

richvdh commented on proposals/1740-accept-header.md:
````diff
@@ -0,0 +1,87 @@
+```
+where accepts is the set of acceptable encodings. This is NOT in `Accept` format because:
+
+1) Servers should not specify a quality or preference, as the negotiations are client led.
````
I don't understand any of these points. Presumably it's not in `Accept` format because the server doesn't know how to format the response in the acceptable format.

richvdh commented on proposals/1740-accept-header.md:
```diff
@@ -0,0 +1,87 @@
+3) Servers know exactly what formats they support and do not need to wildcard the types.
+
+If the clients request contains a `Content-Type` header that the server does not support, 
+the server should respond with `HTTP 415 Unsupported Media Type` and an error of:
```
likewise, why do we bother? is the idea that it provides a mechanism for clients to find an acceptable encoding? I feel like it might be better to do that via a separate mechanism.

Half-Shot replied:
The 415 isn't for finding an encoding, it's for telling the client to stop sending the server stuff it doesn't support at all. For instance: You have a bug in your discovery code.

dbkr commented:
This seems generally quite nice. A couple of things:
 * The C/S spec is currently very JSON-oriented. A bit about how support for other types might be added in the future, ie. the spec gains a section saying how you map from the JSON objects in the spec to new type and back, it it gets added to the list of types a client/server may want to suport.
 * For the same reason, would it make sense to define JSON as a baseline that everyone must support?
 * Presumably clients could make OPTIONS requests as well to discover what types a server will accept? If they get a response saying the server accepts `application/foobar`, can they assume this is accepted for any request they might make to the server? (Strictly I think HTTP would have you query separately per endpoint).

richvdh commented:
> Presumably clients could make OPTIONS requests as well to discover what types a server will accept?

I don't think there is a header defined which does this, so you might as well define a different way.

Half-Shot commented:
I don't think there is much point in trying to define this a different way, the HTTP approach seems sensible. You send the server all the encodings you can support and if it fails, then that's hard luck and you won't be talking to that HS. The responses are purely for consumption for people trying to debug why they cannot connect to their homeserver, but I probably would be fine with dropping it if it's a lot of extra weight.

>Generally, unless we actually plan to implement CBOR-over-HTTP (and given the figures quoted, I'm unconvinced that doing so would buy much over compressed json), this feels a bit like a solution looking for a problem.

I don't think this proposal hinges on the performance of CBOR only, it's designed to just negotiate a format which might also fit other constraints (you might be stuck only being able to talk XML, for instance). The proposal was designed to be useful enough to let server/client implementations experiment with other encodings without having to also write their own negotiation protocols.

As it happens I do have a working set of PRs for Riot/Synapse which probably would be fine to land with some touchups in the next week or so. Given there is talk of already using CBOR over COAP already, we could land the encoding portion while paving the way to support the transport in the future. I'm also going to look at other protocols to see if there is a way to gain more than 20% back, but that certainly seems like a decent improvement for a minor change.

Half-Shot commented:
I'm more convinced that we shouldn't be using JSON for our errors as the only pro I can see is being able to tell a human why their connection failed. Arguably we could probably infer that from the status code, and the best behaved servers might send us some plain text.

I'm in favor of not defining the format of the body, but still enforce the status code.

@turt2live @richvdh @KitsuneRal ?

turt2live commented:
Please let's keep threaded discussions in their threads rather than split them in two places :(

I have replied on the thread already: https://github.com/matrix-org/matrix-doc/pull/1740#discussion_r240329876

maxidor commented:
Just as a reminder, as per [the spec itself](https://matrix.org/docs/spec/client_server/unstable.html#api-standards):
> The mandatory baseline for client-server communication in Matrix is exchanging JSON objects over HTTP APIs.

> For the default HTTP transport, all API calls use a Content-Type of application/json.

> Any errors which occur at the Matrix API level MUST return a "standard error response".

If the MSC could be written keeping those in mind, that would be great. Various parts of the MSC actually go against the spec, or confuse implementation details with spec rules.

Half-Shot commented:
Kewl, that probably covers @dbkr 's concerns then. Since this error occurs at the HTTP level rather than the Matrix level, we don't really need to enforce a standard error response. Other than that, I'm not sure what other "Various" parts go against the spec.

Half-Shot commented:
I forgot this proposal existed, what to do..
